stan/math/prim/mat/fun/log_mix.hpp

```diff
@@ -94,6 +97,274 @@ typename return_type<T_theta, T_lam>::type log_mix(const T_theta& theta,
 
   return ops_partials.build(logp);
 }
+
+/**
+ * Return the log mixture density given specified mixing proportions
+ * and array of log density vectors.
+ *
+ * \f[
+ * \frac{\partial }{\partial p_x}\left[
+ * \log\left(\exp^{\log\left(p_1\right)+d_1}+\cdot\cdot\cdot+
+ * \exp^{\log\left(p_n\right)+d_n}\right)+
+ * \log\left(\exp^{\log\left(p_1\right)+f_1}+\cdot\cdot\cdot+
+ * \exp^{\log\left(p_n\right)+f_n}\right)\right]
+ * =\frac{e^{d_x}}{e^{d_1}p_1+\cdot\cdot\cdot+e^{d_m}p_m}+
+ * \frac{e^{f_x}}{e^{f_1}p_1+\cdot\cdot\cdot+e^{f_m}p_m}
+ * \f]
+ *
+ * \f[
+ * \frac{\partial }{\partial d_x}\left[
+ * \log\left(\exp^{\log\left(p_1\right)+d_1}+\cdot\cdot\cdot+
+ * \exp^{\log\left(p_n\right)+d_n}\right)
+ * +\log\left(\exp^{\log\left(p_1\right)+f_1}+\cdot\cdot\cdot+
+ * \exp^{\log\left(p_n\right)+f_n}\right)\right]
+ * =\frac{e^{d_x}p_x}{e^{d_1}p_1+\cdot\cdot\cdot+e^{d_m}p_m}
+ * \f]
+ *
+ * @param theta vector of mixing proportions in [0, 1].
+ * @param lambda array containing vectors of log densities.
+ * @return log mixture of densities in specified proportion
+ */
+template <typename T_theta, typename T_lam>
+typename return_type<T_theta, std::vector<Eigen::Matrix<T_lam, -1, 1> > >::type
+log_mix(const T_theta& theta,
+        const std::vector<Eigen::Matrix<T_lam, -1, 1> >& lambda) {
+  static const char* function = "log_mix";
+  typedef typename stan::partials_return_type<
+      T_theta, std::vector<Eigen::Matrix<T_lam, -1, 1> > >::type
+      T_partials_return;
+
+  typedef typename Eigen::Matrix<T_partials_return, -1, 1> T_partials_vec;
+
+  typedef typename Eigen::Matrix<T_partials_return, -1, -1> T_partials_mat;
+
+  typedef typename std::vector<Eigen::Matrix<T_lam, -1, 1> > T_lamvec_type;
+
+  const int N = length(lambda);
+  const int M = theta.size();
+
+  check_bounded(function, "theta", theta, 0, 1);
+  check_not_nan(function, "theta", theta);
+  check_finite(function, "theta", theta);
+  for (int n = 0; n < N; ++n) {
+    check_not_nan(function, "lambda", lambda[n]);
+    check_finite(function, "lambda", lambda[n]);
+    check_consistent_sizes(function, "theta", theta, "lambda", lambda[n]);
+  }
+
+  scalar_seq_view<T_theta> theta_vec(theta);
+  T_partials_vec theta_dbl(M);
+  for (int m = 0; m < M; ++m)
+    theta_dbl[m] = value_of(theta_vec[m]);
+
+  T_partials_mat lam_dbl(M, N);
+  vector_seq_view<T_lamvec_type> lam_vec(lambda);
+  for (int n = 0; n < N; ++n)
+    for (int m = 0; m < M; ++m)
+      lam_dbl(m, n) = value_of(lam_vec[n][m]);
+
+  T_partials_mat logp_tmp = log(theta_dbl).replicate(1, N) + lam_dbl;
+
+  T_partials_vec logp(N);
+  for (int n = 0; n < N; ++n)
+    logp[n] = log_sum_exp(logp_tmp.col(n).eval());
+
+  T_partials_mat derivs
```

Comment:
Seems like we only want to calculate the derivatives if we need them - can we put this and the following ops_partials math inside something like `if (!is_constant_struct<T_theta>::value || !is_constant_struct<T_lam>::value)`?

Reply:
That would be a good idea. The matrix ```derivs``` is manipulated in different ways for each input, so would just forgo the calculations if neither input needed derivatives.

i.e.  ```(!(is_constant_struct<T_theta>::value && is_constant_struct<T_lam>::value))```